File: solutions/94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        return_list: List[int] = list()
        stack: List[TreeNode] = list()
        curr: TreeNode = root
        while True:
            if curr:
                stack.append(curr)
                curr = curr.left
            elif len(stack):
                curr = stack.pop()
                return_list.append(curr.val)
                curr = curr.right
            else:
                break
        return return_list
        
        

File: solutions/94. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
  std::vector<int> inorderTraversal(TreeNode* root) {
    auto result = std::vector<int>{};
    auto stack = std::stack<TreeNode*>{};
    TreeNode* curr = root;
    while (curr || !stack.empty()) {
      /* go left if possible */
      while (curr) {
        stack.push(curr);
        curr = curr->left;
      }
      /* leaf node, visit parent and turn to explore the right tree */
      curr = stack.top();
      stack.pop();
      result.push_back(curr->val);
      curr = curr->right;
    }
    return result;
  }
};
​
